x-pack/plugins/security/server/lib/authentication/providers/kerberos.ts

```diff
@@ -0,0 +1,280 @@
+ * document has been removed and ES responds with `500 Internal Server Error`.
+ * @param err Error returned from Elasticsearch.
+ */
+function isAccessTokenExpiredError(err?: any) {
+  const errorStatusCode = getErrorStatusCode(err);
+  return (
+    errorStatusCode === 401 ||
+    (errorStatusCode === 500 &&
+      err &&
+      err.body &&
+      err.body.error &&
+      err.body.error.reason === 'token document is missing and must be present')
+  );
+}
+
+/**
+ * Parses request's `Authorization` HTTP header if present and extracts authentication scheme.
+ * @param request Request instance to extract authentication scheme for.
+ */
+function getRequestAuthenticationScheme(request: Legacy.Request) {
+  const authorization = request.headers.authorization;
+  if (!authorization) {
+    return '';
+  }
+
+  return authorization.split(/\s+/)[0].toLowerCase();
+}
+
+/**
+ * Provider that supports Kerberos request authentication.
+ */
+export class KerberosAuthenticationProvider extends BaseAuthenticationProvider {
+  /**
+   * Performs Kerberos request authentication.
+   * @param request Request instance.
+   * @param [state] Optional state object associated with the provider.
+   */
+  public async authenticate(request: Legacy.Request, state?: ProviderState | null) {
+    this.debug(`Trying to authenticate user request to ${request.url.path}.`);
+
+    const authenticationScheme = getRequestAuthenticationScheme(request);
+    if (authenticationScheme && authenticationScheme !== 'negotiate') {
+      this.debug(`Unsupported authentication scheme: ${authenticationScheme}`);
+      return AuthenticationResult.notHandled();
+    }
+
+    let authenticationResult = await this.authenticateViaHeader(request);
+    if (state && authenticationResult.notHandled()) {
+      authenticationResult = await this.authenticateViaState(request, state);
+      if (authenticationResult.failed() && isAccessTokenExpiredError(authenticationResult.error)) {
+        authenticationResult = AuthenticationResult.notHandled();
+      }
+    }
+
+    // If we couldn't authenticate by means of all methods above, let's try to check if Elasticsearch can
+    // start authentication mechanism negotiation, otherwise just return authentication result we have.
+    return authenticationResult.notHandled()
+      ? await this.authenticateViaSPNEGO(request, state)
+      : authenticationResult;
+  }
+
+  /**
+   * Invalidates access token retrieved in exchange for SPNEGO token if it exists.
+   * @param request Request instance.
+   * @param state State value previously stored by the provider.
+   */
+  public async deauthenticate(request: Legacy.Request, state?: ProviderState) {
+    this.debug(`Trying to deauthenticate user via ${request.url.path}.`);
+
+    if (!state || !state.accessToken) {
+      this.debug('There is no access token invalidate.');
+      return DeauthenticationResult.notHandled();
+    }
+
+    try {
```

Comment:
__note:__ it's probably time to move token handling stuff to a separate shared "module", but not in this PR likely.